R/calcHeatingSystemSales.R

```diff
@@ -0,0 +1,278 @@
+#' countries and European data sets from the EHI and EHPA.
+#'
+#' @returns MagPIE object with heating system sales
+#'
+#' @author Robin Hasse
+#'
+#' @importFrom madrat readSource calcOutput toolGetMapping
+#' @importFrom magclass mselect collapseDim getSets<- complete_magpie
+#' @importFrom dplyr %>% .data filter
+#' @export
+
+calcHeatingSystemSales <- function() {
+
+  # INDIVIDUAL SOURCES ---------------------------------------------------------
+
+  # all heating systems
+  salesTotal <- calcOutput("BRG", aggregate = FALSE, warnNA = FALSE) %>%
+    .removeNADims("region")
+
+  # efficient heating systems (i.e. excl. non-condensing)
+  mapEHI <- toolGetMapping("technologyMapping_EHI.csv",
+                           type = "sectoral", where = "mredgebuildings") %>%
+    filter(.data[["considerSales"]])
+  incompleteHs <- mapEHI %>%
+    filter(!.data[["isComplete"]]) %>%
+    getElement("technologyBRICK") %>%
+    unique()
+  salesEff <- readSource("EHI") %>%
+    mselect(variable = mapEHI[["variable"]]) %>%
+    .removeNADims() %>%
+    .mapToBRICK(mapEHI, from = "variable")
+
+  # heat pumps (used for space heating in buildings)
+  mapEHPA <- toolGetMapping("technologyMapping_EHPA.csv",
+                            type = "sectoral", where = "mredgebuildings") %>%
+    filter(.data[["spaceHeatingHP"]])
+  salesHp <- readSource("EHPA", subtype = "hpSales") %>%
+    mselect(corrected = TRUE) %>%
+    .removeNADims() %>%
+    .mapToBRICK(mapEHPA, from = "hpType")
+
+
+
+  # COMBINE --------------------------------------------------------------------
+
+  ## many European countries ====
+
+  # common regions
+  reg <- Reduce(intersect, lapply(list(salesTotal, salesEff, salesHp),
+                                  getItems, dim = 1))
+
+  # all time steps
+  t <- Reduce(union, lapply(list(salesTotal, salesEff, salesHp),
+                            getItems, dim = 2))
+
+  # extrapolate to all time steps
+  salesTotal <- .time_interpolate(salesTotal, t, 0.2)
+  salesEff <- .time_interpolate(salesEff, t, 0.8)
+  salesHp <- .time_interpolate(salesHp, t, 0.2)
+
+  sales <- mbind(salesEff[reg, , ], salesHp[reg, , ])
+
+  # scale incomplete technologies to match total
+  salesMissing <- salesTotal[reg, , ] - dimSums(sales)
+  salesMissing[salesMissing < 0] <- 0
+  incompleteFactor <- 1 + salesMissing / dimSums(sales[, , incompleteHs])
+  sales[, , incompleteHs] <- sales[, , incompleteHs] * incompleteFactor
+
+  # add zeros for missing technologies
+  sales <- add_columns(sales, c("reel", "sobo"), dim = 3, fill = 0)
+  sales <- add_columns(sales, "dihe", dim = 3, fill = NA)
+
+
+
+  # individual countries ====
+
+  ### AUT ####
+  mapBMK <- toolGetMapping("technologyMapping_BMK.csv",
+                           type = "sectoral",
+                           where = "mredgebuildings")
+  bmk <- readSource("BMK")["AUT", t] %>%
+    .mapToBRICK(mapBMK, from = "technology")
+  hs <- setdiff(getItems(bmk, "hs"), "sobo_biom")
```

Comment:
I would stick to this here, because `hs` actually has (almost) all heating systems that we need in this case. Only `sobo` is missing and calculated in the line below.